backend/Application/Database/MigrationJobs/_03_UpdateFinalizationTimes.cs

```diff
@@ -0,0 +1,161 @@
+                if (state == null)
+                {
+                    return;
+                }
+
+                var initialValues = await FindInitialState(token);
+        
+                var (nextHeight, priorFinalizedBlockHeight, priorFinalizedBlockHash, finalHeight) = initialValues;
+                var initialHeight = nextHeight;
+                while (nextHeight < finalHeight)
+                {
+                    nextHeight++;
+                    _logger.Debug($"{JobName} processing block height {nextHeight}");
+                    var nextBlockInfo = await _client.GetBlockInfoAsync(new Absolute((ulong)nextHeight), token);
+
+                    var nextBlockLastFinalizedHash = nextBlockInfo.BlockLastFinalized.ToString();
+
+                    // Last finalization block hasn't changed
+                    if (nextBlockLastFinalizedHash == priorFinalizedBlockHash)
+                    {
+                        continue;
+                    }
+
+                    await using var context = await _contextFactory.CreateDbContextAsync(token);
+                    var finalizedBlock = await context.Blocks
+                        .SingleAsync(b => b.BlockHash == nextBlockLastFinalizedHash, token);
+
+                    var timeUpdate = new FinalizationTimeUpdate(priorFinalizedBlockHeight, finalizedBlock.BlockHeight);
+                    _logger.Debug($"Updating block height from:{timeUpdate.MinBlockHeight}, to {timeUpdate.MaxBlockHeight}");
+                    
+                    await BlockWriter.UpdateFinalizationTimes(timeUpdate, nextBlockInfo.BlockSlotTime, context);
+                    await _metricsWriter.UpdateFinalizationTimes(timeUpdate);
+
+                    priorFinalizedBlockHash = finalizedBlock.BlockHash;
+                    priorFinalizedBlockHeight = finalizedBlock.BlockHeight;
+                    ApplicationMetrics.SetJobCompletion(this, (nextHeight - initialHeight) / (double)finalHeight);
+                }
+            });
+        _logger.Debug($"Done processing {JobName}");
+    }
+
+    private sealed record InitialState(
+        int NextHeight,
+        int PriorFinalizedBlockHeight,
+        string PriorFinalizedBlockHash,
+        int FinalHeight
+    );
+
+    private async Task<InitialState> FindInitialState(CancellationToken token)
+    {
+        await using var context = await _contextFactory.CreateDbContextAsync(token);
+
+        var firstBlockWithNull = await context
+            .Blocks
+            .Where(b => b.BlockStatistics.FinalizationTime == null)
+            .OrderBy(b => b.BlockHeight)
+            .Take(1)
+            .SingleAsync(cancellationToken: token);
+            
+        var priorBlockInfo = await _client.GetBlockInfoAsync(new Absolute((ulong)firstBlockWithNull.BlockHeight - 1), token);
+        var priorBlockFinalization = await context.Blocks
+            .SingleAsync(x => x.BlockHash == priorBlockInfo.BlockLastFinalized.ToString(), cancellationToken: token);
```

Comment:
I am not sure why we are using this. `firstBlockWithNull` is the first block that isn't recorded as being finalized by another block, so the parent of that is recorded as being finalized by another block. We shouldn't need to consider the last finalized block with respect to that block, which is what we're doing here.

In particular, I think the loop will behave incorrectly on the first iteration with this definition, because the parent of the `firstBlockWithNull` will be changed to being finalized by the finalizer of `firstBlockWithNull`, though its finalization time is already recorded.